addons/controllers/kapp-controller/kappcontrollerconfig_controller.go

```diff
@@ -0,0 +1,258 @@
+// move the current state of the cluster closer to the desired state.
+func (r *KappControllerConfigReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("kappControllerConfig", req.NamespacedName)
+
+	// get kapp-controller config object
+	kappControllerConfig := &runv1alpha3.KappControllerConfig{}
+	if err := r.Client.Get(ctx, req.NamespacedName, kappControllerConfig); err != nil {
+		log.Error(err, "unable to fetch kappControllerConfig")
+		return ctrl.Result{}, err
+	}
+
+	// get the parent cluster name from owner reference
+	// if the owner reference doesn't exist, use the same name as config CRD
+	clusterNamespacedName := req.NamespacedName
+	cluster := &clusterapiv1beta1.Cluster{}
+	for _, owner := range kappControllerConfig.OwnerReferences {
+		if owner.Kind == cluster.Kind {
+			clusterNamespacedName.Name = owner.Name
+			break
+		}
+	}
+
+	// verify that the cluster related to config is present
+	if err := r.Client.Get(ctx, clusterNamespacedName, cluster); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("Cluster not found")
+			return ctrl.Result{}, nil
+		}
+
+		log.Error(err, "unable to fetch cluster")
+		return ctrl.Result{}, err
+	}
+
+	if retResult, err := r.ReconcileKappControllerConfig(ctx, log, cluster, kappControllerConfig); err != nil {
+		log.Error(err, "unable to reconcile kappControllerConfig")
+		return retResult, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *KappControllerConfigReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, options controller.Options) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&runv1alpha3.KappControllerConfig{}).
+		WithOptions(options).
+		Complete(r)
+}
+
+// ReconcileKappControllerConfig reconciles KappControllerConfig CRD
+func (r *KappControllerConfigReconciler) ReconcileKappControllerConfig(
+	ctx context.Context,
+	log logr.Logger,
+	cluster *clusterapiv1beta1.Cluster,
+	kappControllerConfig *runv1alpha3.KappControllerConfig) (_ ctrl.Result, retErr error) {
+
+	var (
+		patchConfig bool
+	)
+
+	patchHelper, err := clusterapipatchutil.NewHelper(kappControllerConfig, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	// Patch KappControllerConfig before returning the function
+	defer func() {
+		// patchConfig will be true if finalizer, ownerrefence or secretRef is added or deleted
+		if patchConfig {
+			log.Info("Patching kappControllerConfig")
+
+			if err := patchHelper.Patch(ctx, kappControllerConfig.DeepCopy()); err != nil {
+				log.Error(err, "Error patching kappControllerConfig")
+				retErr = err
+			}
+		}
+	}()
+
+	// If KappControllerConfig is marked for deletion then delete the data value secret
+	if !kappControllerConfig.GetDeletionTimestamp().IsZero() {
+		log.Info("Deleting kappControllerConfig")
+		err := r.ReconcileKappControllerConfigDelete(ctx, log, cluster, kappControllerConfig, &patchConfig)
```

Comment:
thought: I don't think deletion should trigger secret delete because that would make it hard to remove the package on the cluster. The secret will be automatically deleted when ownerRefs are gone and deletion for kappconfig shouldn't do anything except not attempting to do any reconcile.

Reply:
The deletion here doesn't delete the config CR, it deletes the data value secret the controller created.

Reply:
@vijaykatam Who will be the owners to the secrets in this case? Is the package the owner?

Reply:
deletion is a noop now